lib/Mojo/IOLoop/Stream/HTTPClient.pm

```diff
@@ -0,0 +1,202 @@
+}
+
+sub _read_content {
+  my ($self, $chunk) = @_;
+
+  # Corrupted connection
+  return $self->close unless my $tx = $self->{tx};
+
+  warn term_escape "-- Client <<< Server (@{[_url($tx)]})\n$chunk\n" if DEBUG;
+  $tx->client_read($chunk);
+  $self->_finish if $tx->is_finished;
+}
+
+sub _upgrade {
+  my ($self, $tx) = @_;
+  my $code = $tx->res->code // 0;
+  return undef unless $tx->req->is_handshake && $code == 101;
+  my $ws = Mojo::Transaction::WebSocket->new(handshake => $tx, masked => 1);
+  return challenge($ws) ? $ws->established(1) : undef;
+}
+
+sub _url { shift->req->url->to_abs }
+
+sub _write_content {
+  my $self = shift;
+
+  # Protect from resume event recursion
+  return if !(my $tx = $self->{tx}) || $self->{cont_writing};
+  local $self->{cont_writing} = 1;
+  my $chunk = $tx->client_write;
+  warn term_escape "-- Client >>> Server (@{[_url($tx)]})\n$chunk\n" if DEBUG;
+  return unless length $chunk;
+  $self->write($chunk => sub { $_[0]->_write_content });
+}
+
+1;
+
+=encoding utf8
+
+=head1 NAME
+
+Mojo::IOLoop::Stream::HTTPClient - Non-blocking I/O HTTP client
+
+=head1 SYNOPSIS
+
+  use Mojo::IOLoop::Client;
+  use Mojo::IOLoop::Stream::HTTPClient;
+  use Mojo::Transaction::HTTP;
+  
+  # Create transaction
+  my $tx = Mojo::Transaction::HTTP->new;
+  $tx->req->method('GET')->url->parse('http://mojolicious.org');
+  $tx->on(finish => sub {
+      my $tx = shift;
+      say $tx->res->code;
+    });
+  
+  # Create socket connection
+  my $client = Mojo::IOLoop::Client->new;
+  $client->on(
+    connect => sub {
+      my $stream = Mojo::IOLoop::Stream::HTTPClient->new(pop);
+      $stream->start;
+      $stream->process($tx);
+    }
+  );
+  
+  $client->connect(address => 'mojolicious.org', port => 80);
+  $client->reactor->start;
+
+=head1 DESCRIPTION
+
+HTTP client connection based on L<Mojo::IOLoop::Stream>.
+
+=head1 EVENTS
+
+L<Mojo::IOLoop::Stream::HTTPClient> inherits all events from
+L<Mojo::IOLoop::Stream::Client> and can emit the following new ones.
+
+=head2 upgrade
+
+  $stream->on(upgrade => sub {
+    my ($stream, $ws) = @_;
+    ...
+  });
+
+Emitted when connection should be upgraded to WebSocket.
+
+=head1 ATTRIBUTES
+
+L<Mojo::IOLoop::Stream::HTTPClient> inherits all attributes from
+L<Mojo::IOLoop::Stream::Client> and implements the following ones.
+
+=head2 request_timeout
+
+  my $timeout = $stream->request_timeout;
+  $stream     = $stream->request_timeout(5);
+
+Maximum amount of time in seconds sending the request and receiving a whole
+response may take before getting canceled, defaults to the value of the
+C<MOJO_REQUEST_TIMEOUT> environment variable or C<0>. Setting the value to C<0>
+will allow to wait indefinitely.
+
+=head1 METHODS
+
+L<Mojo::IOLoop::Stream::HTTPClient> inherits all methods from
+L<Mojo::IOLoop::Stream::Client> and implements the following new ones.
+
+=head2 process
+
+  $stream->process(Mojo::Transaction::HTTP->new);
+
+Process a transaction with current connection.
```

Comment:
Maybe `Process a L<Mojo::Transaction::HTTP> object with the current connection`.